stdlib/Random/src/misc.jl

```diff
@@ -183,6 +183,28 @@ ltm52(n::Int, mask::Int=nextpow(2, n)-1) = LessThan(n-1, Masked(mask, UInt52Raw(
 
 ## shuffle & shuffle!
 
+function shuffle(rng::AbstractRNG, tup::NTuple{N}) where {N}
+    @inline let  # `@inline` and `@inbounds` are here to help escape analysis
+        # use a narrow integer type to save stack space and prevent heap allocation
+        Ind = if N ≤ typemax(UInt8)
+            UInt8
+        elseif N ≤ typemax(UInt16)
+            UInt16
+        else
+            UInt
+        end
+        clo = @inbounds let mem = Memory{Ind}(undef, N)
+            shuffle!(rng, copyto!(mem, Base.OneTo(N)))  # just use `randperm!` once it supports `Memory`
+            let mem = mem, tup = tup
```

Comment:
I tested both with and without this `let` and couldn't notice a difference. Is it really necessary here?

Reply:
I see equal (or even little better) performance for simple
```julia
function shuffle(rng::AbstractRNG, tup::NTuple{N}) where {N}
    Ind = if N ≤ typemax(UInt8)
        UInt8
    elseif N ≤ typemax(UInt16)
        UInt16
    else
        UInt
    end

    mem = Memory{Ind}(undef, N)
    shuffle!(rng, copyto!(mem, Base.OneTo(N)))
    return ntuple(i -> tup[mem[i]], Val(N))
end
```